diff --git a/node_modules/@maplibre/mlt/dist/vector/geometry/gpuVector.js b/node_modules/@maplibre/mlt/dist/vector/geometry/gpuVector.js
index 3b9eda4..1be0afd 100644
--- a/node_modules/@maplibre/mlt/dist/vector/geometry/gpuVector.js
+++ b/node_modules/@maplibre/mlt/dist/vector/geometry/gpuVector.js
@@ -97,6 +97,42 @@ export class GpuVector {
                         geometries[i] = allRings;
                     }
                     break;
+                case GEOMETRY_TYPE.LINESTRING:
+                    {
+                        partOffsetCounter++;
+                        const numVertices = ringOffsets[ringOffsetsCounter] - ringOffsets[ringOffsetsCounter - 1];
+                        ringOffsetsCounter++;
+                        const line = [];
+                        for (let k = 0; k < numVertices; k++) {
+                            const x = this._vertexBuffer[vertexBufferOffset++];
+                            const y = this._vertexBuffer[vertexBufferOffset++];
+                            line.push(new Point(x, y));
+                        }
+                        geometries[i] = [line];
+                        if (geometryOffsets)
+                            geometryOffsetsCounter++;
+                    }
+                    break;
+                case GEOMETRY_TYPE.MULTILINESTRING:
+                    {
+                        const numLineStrings = geometryOffsets[geometryOffsetsCounter] - geometryOffsets[geometryOffsetsCounter - 1];
+                        geometryOffsetsCounter++;
+                        const lineStrings = [];
+                        for (let j = 0; j < numLineStrings; j++) {
+                            partOffsetCounter++;
+                            const numVertices = ringOffsets[ringOffsetsCounter] - ringOffsets[ringOffsetsCounter - 1];
+                            ringOffsetsCounter++;
+                            const line = [];
+                            for (let k = 0; k < numVertices; k++) {
+                                const x = this._vertexBuffer[vertexBufferOffset++];
+                                const y = this._vertexBuffer[vertexBufferOffset++];
+                                line.push(new Point(x, y));
+                            }
+                            lineStrings.push(line);
+                        }
+                        geometries[i] = lineStrings;
+                    }
+                    break;
             }
         }
         return geometries;
